### ISSUE TITLE
Support for dtypes

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy_stubs#7]
Per https://github.com/numpy/numpy/issues/7370#issuecomment-346126667, we could specify dtype as part of ndarray types (including type inference) by using NumPy's [scalar types](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.scalars.html), e.g., `np.ndarray[Any, np.float64]` or `np.ndarray[np.float64, Any]` (where `Any` indicates the shape).

Is this a good way to indicate dtypes? How would we handle dtypes that don't have a well specified scalar type? For example:
- `datetime64` and `timedelta64` do not have precision as part of the type
- [`np.void`](https://stackoverflow.com/questions/25247190/numpy-void-type-how-to-use-it) is used for all arbitrary structured dtypes, but doesn't indicate the field names or subtypes (side note: `np.struct` would be a much better name)
- string dtypes have their size as part of the type

One simple answer is to not worry about more detailed type information for now, and worry about more precise specification for scalar types later, when support in the typing module becomes available (e.g., for integer types or custom struct-like data structures like `TypedDict`).

**Comment by rmcgibbo** [comment]
I think `ndarray[ShapeType, DtypeType]` is the best way to so specify the type.

My reason for preferring that the shape specification come before the dtype specification is that I anticipate that when things get a little verbose, we'll probably want to have type aliases, and putting the `ShapeType` first in the syntax makes something like a `array1d[Dtype]` as a shorthand (that looks like a partially applied function) for `ndarray["OneDimensional", Dtype]` makes much more sense than `arrayf64[ShapeType]` as a shorthand for `ndarray[ShapeType, float64]`.

**Comment by rmcgibbo** [comment]
For  complex dtypes like `datetime64`, `np.void`, and the string dtypes, I propose that we delay dealing with them until this package is much more fully developed.

Very little functionality is lost in the meantime if numpy functions that return arrays of `dtype('S1')` are labeled as returning (or are inferred to return) results of type `ndarray[Any, str]`. More specificity can be added later.

**Comment by ryanpeach** [comment]
This is really all I need, what is the issue? Anything I can do?

**Comment by person142** [comment]
Circling back around to this: another issue is that you can mutate an array's dtype in-place, e.g.

```python
>>> x = np.ones((2,), dtype=np.int64)
>>> x
array([1, 1])
>>> x.dtype = np.bool_
>>> x
array([ True, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False, False])
```

So if you tried to add types:

```python
x: np.ndarray[np.int64] = np.ones((2,), dtype=np.int64)
x.dtype = np.bool_  # Uh oh, the type is incorrect now.
```

I imagine that most of the time people don't do that (and use views instead), which we could handle quite nicely in some cases with the right overload:

```python
T = TypeVar('T', bound=dtype)

class ndarray(...):
    ...
    @overload
    def view(self, dtype: T = ...) -> ndarray[T]: ...
```

A practical solution might be to say "if you are going to set dtype, you are on your own", and ask that in such cases a user do

```python
x: np.ndarray[Any] = np.ones((2,), dtype=np.int64)
x.dtype = np.bool_  # Still ok
```

Thoughts?

**Comment by shoyer** [comment]
I don't think we should support modifying array `dtype` in place. There's really not good reason to do this rather than making a new view.

Likewise, I don't think we should support mutating `shape`, though if I recall that may be the only way to do reshaping while ensuring that an error would be raised if a copy would be needed.

**Comment by person142** [comment]
> I don't think we should support modifying array dtype in place

Should we go so far as removing the setter from the types?

https://github.com/numpy/numpy-stubs/blob/master/numpy-stubs/__init__.pyi#L300

**Comment by person142** [comment]
PR banning the setter: https://github.com/numpy/numpy-stubs/pull/47
